microcosm_postgres/tests/test_employee_store_with_encryption.py

```diff
@@ -0,0 +1,276 @@
+            retrieved_employee = retrieved_employees[0]
+            assert retrieved_employee.name == "foo"
+            assert retrieved_employee.id == employee.id
+
+
+def test_encrypt_no_beacon_used(
+    session: Session,
+    single_tenant_encryptor: SingleTenantEncryptor,
+    graph: ObjectGraph,
+) -> None:
+    with AwsKmsEncryptor.set_encryptor_context("test", single_tenant_encryptor):
+        session.add(employee := Employee(name="foo", salary=100))
+        assert employee.salary_unencrypted is None
+        assert employee.salary_encrypted is not None
+        with pytest.raises(AttributeError):
+            assert employee.salary_beacon is None  # type: ignore[attr-defined]
+
+        assert employee.name == "foo"
+        assert employee.salary == 100
+
+
+def test_order_by_with_beacon(
+    single_tenant_encryptor: SingleTenantEncryptor,
+    graph: ObjectGraph,
+) -> None:
+    """
+    Test that when we order by with the beaconised field, then it
+    doesn't error.
+
+    Note that if you try to order with a beaconised field it will
+    be essentially random.
+
+    """
+    with SessionContext(graph) as context:
+        context.recreate_all()
+
+        session = context.session
+        with AwsKmsEncryptor.set_encryptor_context("test", single_tenant_encryptor):
+            session.add(Employee(name="foo", salary=1000))
+            session.add(Employee(name="bar", salary=1000))
+            session.commit()
+
+    with AwsKmsEncryptor.set_encryptor_context("test", single_tenant_encryptor):
+        query = select(Employee).order_by(Employee.name.asc())
+        # Use regex to match the compiled sql
+        regex = re.compile(r"ORDER BY .*?name_beacon ASC")
+        assert regex.search(str(query))
+
+        results = session.execute(query).scalars().all()
+        assert len(results) == 2
+        for r in results:
+            assert r.salary == 1000
+
+        # Same for desc
+        query = select(Employee).order_by(Employee.name.desc())
+        regex = re.compile(r"ORDER BY .*?name_beacon DESC")
+        assert regex.search(str(query))
+
+        results = session.execute(query).scalars().all()
+        assert len(results) == 2
+        for r in results:
+            assert r.salary == 1000
+
+
+def test_searching_on_encrypted_field_with_no_beacon(
+    session: Session,
+    single_tenant_encryptor: SingleTenantEncryptor,
+    graph: ObjectGraph,
+):
+    """
+    This test checks that it doesn't break if you try to search for a field
+    which is encrypted but with no beacon field defined.
+
+    """
+
+    with AwsKmsEncryptor.set_encryptor_context("test", single_tenant_encryptor):
+        session.add(Employee(name="foo", salary=1000))
+        session.add(Employee(name="bar", salary=1000))
+        session.commit()
+
+    query = select(Employee).filter(Employee.salary == 1000).order_by(Employee.name.asc())
+    results = session.execute(query).scalars().all()
+
+    assert len(results) == 0
+
```

Comment:
I think you need a test where we add employee(s) without encryption and then we do the search

I'm not sure it works in this case?

Reply:
True, it didn't work. I added a test case to cover and updated the implementation.